### ISSUE TITLE
[Firestore] openDBChannel

### ISSUE THREAD

**Issue by louisameline** [vuex-cloud/rfcs#13]
I'll drop this for later use, I know it's not crystal clear yet.

Two things:

1. The lib currently closes the channel if the document doesn't exist or which gets deleted. But actually it's theoretically ok to listen to do so, even without write rights. The user could want to keep listening for a later insert, and we should allow that.

Also, in such case he may want to unsubscribe but keep the module active, or unsubscribe and destroy the module automatically if it was dynamically registered. We'll need to make this easy.

2. The initial promise sent by the `openDBChannel` action is kind of awkward, because the stream is actually effective right away (synchronously). The promise is currently used to tell users that their data has been synchronized with Firestore, in the perspective that the app couldn't be used before that, at the risk of losing data when connectivity is recovered. But actually this doesn't apply to all use cases, and one might want to start using the app and use a strategy of merging data when going online, which is not straightforward right now.
Basically I think the promise of the action should be the `streaming` promise which got implement in the latest version, and the "it's alright, we received data from Firestore" information should be made available another way, possibly with events.

**Comment by louisameline** [comment]
Hi Luca, I update my post above. I just noticed you hadn't published the latest version of the documention... and that's very convenient! It means nobody knows about the promise now returned by `openDBChannel`, I still can change that to return an object instead to implement the signature above :D I'll update my PR again now!

**Comment by mesqueeb** [comment]
Ok ;) I hope I can merge it tomorrow ! Thanks a lot for all your hard work.

**Comment by louisameline** [comment]
You're welcome, you've been working pretty hard on it too ;)

So, `refreshed` and `stop` are now implemented as suggested above.
For  `refreshed`, I had to implement an option to pass to the action, because it required to enable `includeMetadataChanges`, which is quite special and won't be suited for everybody.

So I edited the syntax of the action to be 
```javascript
dispatch('module/openDBChannel', {pathVariables: {}, includeMetadataChanges: true})
```
It's all backwards compatible with the old syntax, no breaking changes.

Only afterwards did I notice that you also had a similar syntax on the `fetch` actions where the parameter is a mix of user path variables and clauses (where, orderBy, limit). But this was already dreadful to me, I wasn't going to add options in there too x)

So I really needed to fix this and make the `fetch*` actions consistent with the new syntax of `openDBChannel`. So they're now like
```javascript
dispatch('module/fetchAndAdd', {clauses: {where, orderBy, limit}, pathVariables: {}, ...options})
```
They too are backward compatible with the old syntax.
By the way, since `limit` and `orderBy` are not "filters", I renamed "filter" by "clause" everywhere.
I've edited all the documentation accordingly, maybe you should start there when you review the changes.
Thanks!

**Comment by louisameline** [comment]
I updated the initial post with what I think we should do in the next major version with `openDBChannel` (or `stream`, whatever the name). I know you've already posted about the events we need to take into consideration in the lib, I need some time later in the month to read it and reply.

**Comment by mesqueeb** [comment]
@louisameline I left you some messages on discord.

here's the TLDR;

(unrelated of calling the module instance directly or via vuex)
instead of:
```js
myModule.stream()
  .then(() => {
    // this gets resolved when you close the channel yourself
  })
  .catch(error => {
    // an error occured at some point that made the channel close
  })
```

I'm thinking more like:
```js
myModule.stream()
  .then(unsubscribeFunction => {
    // this gets resolved immediately when your plugins (eg. Firebase) succesfully opened the stream
    // `unsubscribeFunction` can be executed to close the stream
  })
  .catch(error => {
    // an error occured at some point WHILE OPENING THE stream
  })

// and to catch edge cases you could do this as well:
myModule.stream({errorHandler})
```

**Comment by louisameline** [comment]
Reminder: when `stream` is called while the same stream was already open, it's good to resolve the promise, but we'd want to include an `alreadyOpen` flag in the resolved value, so the user may skip any steps he has already done at the previous stream opening.
